### PR TITLE
Dash oncoprint

### PR DESCRIPTION
**Pull request by jackluo**
# Dash Bio pull request
- [x] This is a new component 
- [ ] I am adding a feature to an existing component
- [ ] I'm improving an existing feature

## Name and description of your component
Dash oncoprint - heatmaps for genomic alterations

## Main libraries used in component
* Make sure you add links!

## Before asking for a review
- [x] I have gone through the [code review checklist](https://github.com/plotly/dash-component-boilerplate/blob/master/%7B%7Bcookiecutter.project_shortname%7D%7D/review_checklist.md)

## PR merging checklist
Please make sure you have done these things before asking for approval to finally merge. 
- [x] Address all comments in the code review.
- [x] Make sure you are up-to-date with the latest `master`: check out the master branch with `git checkout master`, pull the latest changes with `git pull origin master`, check out back to your branch with `git checkout your-branch`, merge (latest) master into your branch with `git merge master`. [Resolve conflicts](https://help.github.com/articles/resolving-a-merge-conflict-using-the-command-line/) if there are any, and finally `git push origin your-branch`.
- [x] Update/regenerate `package-lock.json` (run `npm install`)  
- [x] Ensure that your `requirements.txt` is complete
- [x] Rebuild (run `npm run build:all` or `npm run build:js` followed by `npm run build:py`
- [x] Regenerate tarball (run `python setup.py sdist`, then copy `dist/dash_bio-X.Y.Z.tar.gz` into the top-level directory)
- [x] Ensure that you installed the tarball you generated with the above commands (run `pip install dash-bio-X.Y.Z.tar.gz`)
- [x] Run your application successfully in a [virtual environment](https://realpython.com/python-virtual-environments-a-primer/) and test that other apps still run
- [x] Commit and push the files that were generated in the `dash_bio/` subfolder by the above command(s)
- [x] Ask for a final review and look for a :dancer:!
- [ ] Have your dancer? Merge to master!

## DDS Deployment
- [x] Get the `developers` user credentials from a dash-bio team member.
- [x] Checkout the master branch of the dash-bio repository, make sure you have the latest master (`git pull`).
- [x] Check remotes associated to your repository: `git remote -v`, make sure you have the dash-bio one: `https://dash-gallery.plotly.host/GIT/dash-bio`.
- [x] If you don't have the dash-bio remote add it with https, ssh will not work: `git remote add dash-bio https://dash-gallery.plotly.host/GIT/dash-bio`.
- [x] You will be asked to enter the `developers` user credentials, have them ready.
- [x] Push local master to the dds dash-bio master: `git push dash-bio master`.

### DDS Deployment Debugging

- If running `git push dash-bio master` you may encounter the following issue:
```
remote: User <username> does not have permissions to run git-hook on dash-bio, or dash-bio does not exist
remote: Access denied
To https://dash-gallery.plotly.host/GIT/dash-bio
 ! [remote rejected] master -> master (pre-receive hook declined)
error: failed to push some refs to 'https://dash-gallery.plotly.host/GIT/dash-bio'
```

This may be because the `git` credentials are saved on the `client-side`. In order to solve this, do the following: 

#### Windows 
```
Go to: Control Panel -> User Accounts -> Manage your credentials -> Windows Credentials -> Under Generic Credentials find: git:https://dash-gallery.plotly.host -> Remove credential OR edit username/password to provided account.
```

#### OS X 
```
Go to: Keychain Access -> Search for dash-gallery.plotly.host -> Change "Account" to "developers" -> Click on "show password" -> Change the password to the one provided -> Click "Save Changes". 
```

### CODE REVIEW

File: src/lib/components/OncoPrint.react.js [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+    /**
+     * If not null, will override the default OncoPrint colorscale.
+     * Default OncoPrint colorscale same as CBioPortal implementation.
+     * Make your own colrscale as a {'mutation': COLOR} dict.
+     * Supported mutation keys in ['MISSENSE, 'INFRAME', 'FUSION',
+     * 'AMP', 'GAIN', 'HETLOSS', 'HMODEL', 'UP', 'DOWN']
+     * Note that this is NOT a standard plotly colorscale.
+     */
+    colorscale: PropTypes.oneOfType([
+        PropTypes.bool,
```
[the remark this answers is not in the document]
Comment (jackluo) [reply src/lib/components/OncoPrint.react.js]: used to accept null

File: src/lib/components/OncoPrint.react.js [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+     * Width of the OncoPrint.
+     * Will disable auto-resizing of plots if set.
+     */
+    width: PropTypes.oneOfType([
+        PropTypes.number,
+        PropTypes.string
+    ]),
+
+    /**
+     * Width of the OncoPrint.
```
[the remark this answers is not in the document]
Comment (jackluo) [reply src/lib/components/OncoPrint.react.js]: fixed

File: src/lib/components/OncoPrint.react.js [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+     *.Toogles whether or not to show a heatmap overview of the tracks.
+     */
+    showoverview: PropTypes.bool,
+
+    /**
+     * Width of the OncoPrint.
+     * Will disable auto-resizing of plots if set.
+     */
+    width: PropTypes.oneOfType([
+        PropTypes.number,
```
[the remark this answers is not in the document]
Comment (jackluo) [reply src/lib/components/OncoPrint.react.js]: to keep consistent with default plotly.py behaviour
most number props in the python api accept them as strings as well

File: tests/dash/app_oncoprint.py [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+                                    ),
+                                    dcc.Textarea(
+                                        id="oncoprint-events",
+                                        placeholder="Hover or click on data to see it here.",
+                                        value="Hover or click on data to see it here.",
+                                        className="oncoprint-events",
+                                    ),
+                                ], className='oncoprint-subcard'),
+                                html.Div([
+                                    html.H4(
```
[the remark this answers is not in the document]
Comment (jackluo) [reply tests/dash/app_oncoprint.py]: thanks!

**Comment by jackluo** [comment]
All done - let's merge asap

**Comment by shammamah** [comment]
Awesome work! Just a few things to address: 

screen shot 2019-01-28 at 4 39 38 pm

Let's get rid of the `<br>` here (if it would require a big change, we can save it for a subsequent PR). 

Could you also please pick a different color for the header? It's just a bit difficult to see the logo. 

Other than that, it looks amazing :)

**Comment by jackluo** [comment]
Can easily fix the <br> thing on the React component which is in a separate repo - not sure if its the right thing to do though:

The reason why <br> is inserted is to hack a multiline tooltip (afaik there's no way otherwise)
In either case, it's an anti-pattern to reference the text itself on a hover event, it is mostly there just for debugging/convenience purposes. Crossfilterering should ideally always be on curveNumber 

Lmk thoughts
Fixed header color  :)

**Comment by shammamah** [comment]
@jackluo Might be a good idea to let the user figure out how to add a multiline tooltip. I think people might want to parse the JSON hover/click data that you have, and a `<br>` in the middle might make that difficult. 

But we can leave that for later -- looks great for now! 💃

**Comment by jackluo** [comment]
Merged

**Pull request merged**